src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -680,207 +680,197 @@ class DistributedClosestPointImpl
       isBVHTreeInitialized(),
       "BVH tree must be initialized before calling 'computeClosestPoints");
 
-    BoxType queryPartitionBb = computeMeshBoundingBox(queryMesh, coordset);
+    BoxType myQueryBb = computeMeshBoundingBox(queryMesh, coordset);
+    BoxArray allQueryBbs;
+    gatherBoundingBoxes(myQueryBb, allQueryBbs);
 
     std::map<int, std::shared_ptr<conduit::Node>> xferNodes;
-    std::list<std::shared_ptr<conduit::Node>> skipNodes;
 
     // create conduit node containing data that has to xfer between ranks.
     // The node will be mostly empty if there are no query points on this rank
     {
       xferNodes[m_rank] = std::make_shared<conduit::Node>();
       conduit::Node& xferNode = *xferNodes[m_rank];
       copy_query_node_to_xfer_node(queryMesh, xferNode, coordset);
-      put_bounding_box_to_conduit_node(queryPartitionBb, xferNode["aabb"]);
+      put_bounding_box_to_conduit_node(myQueryBb, xferNode["aabb"]);
+      xferNode["homeRank"] = m_rank;
+    }
+
+    {
+      conduit::Node& xferNode = *xferNodes[m_rank];
+      computeLocalClosestPointsByPolicy(xferNode);
+    }
 
-      if(m_isVerbose)
+    const auto& myObjectBb = m_objectPartitionBbs[m_rank];
+    int remainingRecvs = 0;
+    for(int r = 0; r < m_nranks; ++r)
+    {
+      if(r != m_rank)
       {
-        internal::dump_node(xferNode,
-                            fmt::format("round_{}_r{}_begin.json", 0, m_rank));
+        const auto& otherQueryBb = allQueryBbs[r];
+        double sqDistance =
+          axom::primal::squared_distance(otherQueryBb, myObjectBb);
+        if(sqDistance <= m_sqDistanceThreshold)
+        {
+          ++remainingRecvs;
+        }
       }
     }
 
-    /*
-      Send query partition to other processes for searches.  When it
-      returns, search against local object partition.  Query partitions
-      are sent in a ring pattern.  If an object partition is farther
-      than the threshold distance from the query partition, skip over
-      that rank.
-    */
-
     // arbitrary tags for send/recv xferNode.
     const int tag = 987342;
 
     std::list<conduit::relay::mpi::Request> isendRequests;
 
-    int toRecvCount = m_nranks - 1;  // Expect data for each non-local process.
-
-    auto xferNodePtr = xferNodes.at(m_rank);
-    for(int round = 0; round < m_nranks || xferNodePtr; ++round)
     {
-      SLIC_INFO_IF(
-        m_isVerbose,
-        fmt::format("=======  Starting round {}/{} =======", round, m_nranks));
-
-      if(xferNodePtr && m_nranks > 1)
+      /*
+        Send local query mesh to next rank with close-enough object
+        partition, if any.  Increase remainingRecvs, because this data
+        will come back.
+      */
+      int firstRecipForMyQuery = next_recipient(*xferNodes[m_rank]);
```

Comment:
Do the queries have to go in ring order? Or is this just a convenient way to order them?

I was wondering if it would make sense to sort the recipients by increasing (max) distance that way we're likely to get a pretty good approximation in the first few rounds, and might be able to skip (or at least return early) from the later rounds.

(max distance == the max possible distance between the bounding box of a query mesh and the bounding box of an object mesh)

Reply:
The ring order is a convenient way to order them and has the advantage that the sender and receiver, being close in rank-space, are likely to be close on the physical network.  Sorting by increasing distance will scramble this order.  I don't think the early-stop will be more performant because sorting requires an O(N) or O(N logN) algorithm, which is no faster than not trying to stop early.

Reply:
True, but I was thinking that `O(N log N)` _on rank_ (where N is hundreds or even thousands) is probably faster than an `~O(1)` algorithm that has to be run `N` times, and each of those might have additional MPI comms.

I think the difference might be that the algorithm I was suggesting would account for the max distance among the object bounding boxes (and possibly also the local approximation to the distance) and not just the distance threshold.

Reply:
I think a main use-case for the distance threshold is that the end-user/application only needs to process items that are withing the given threshold and can safely ignore farther distances.

Reply:
So you're saying that with threshold and without are two distinct use cases.  If so, then sorting would buy us **a lot** in the latter case.